library/src/main/java/com/getstream/sdk/chat/ChatImpl.java

```diff
@@ -109,8 +100,23 @@ public String getVersion() {
         return BuildConfig.BUILD_TYPE + ":" + BuildConfig.VERSION_NAME;
     }
 
-    void init() {
+    @Override
+    public void setUser(@NotNull User user, @NotNull String token, @NotNull InitConnectionListener callbacks) {
+        ChatClient.instance().setUser(user, token, new InitConnectionListener() {
+            @Override
+            public void onSuccess(@NotNull ConnectionData data) {
+                ChatDomain.instance().setCurrentUser(user);
```

Comment:
Why `ChatDomain` is mandatory on the Builder, but it is not injected into `Chat` instance?

Reply:
Since `ChatDomain` like `ChatClient` is designed to be public, immutable singletons I don't think there's any reason to store some more local references to the global one.

Anyway, I find it a bit weird because having those singletons `Chat.Builder` class acts rather as a validator and checks if `ChatClient` and `ChatDomain` are initialized properly. 

Since `Chat.setUser()` was supposed to be main action/entry point, maybe it would be better to validate initialization of `ChatClient` and `ChatDomain` inside it? 

Imo, this would be cleaner not to force passing `ChatClient` and `ChatDomain` to builder. But could be a bit misleading for the user. 

1. Any ideas for where to do the validation? 
2. Do you think we should force users to pass `ChatClient` and `ChatDomain` to builder just for validation purposes? 

This would require the user to perform the following operations in the correct order.
```
1. ChatClient.Builder().build()
2. ChatDomain.Builder().build()
3. Chat.Builder().build()
```
So this "correct order" requirement could be errorgenous. 

Another way would be to hide the `ChatClient` and `ChatDomain` initialization completely from the user and wrap it by `Chat.Builder().build()` or `Chat.setUser()`.

Calling for ideas here, @tschellenbach @elevenetc @tbarbugli

Reply:
Build everything at once without required `client` and `domain`, because `setUser` might be called later or never.
```kotlin
val chat = Chat.Builder(apiKey).build()
//much later or never
chat.setUser(user)
```

Reply:
Some points:
-. Our dependencies are "kind of singleton" but they are not immutable, whenever you call `build()` method, the instance is mutated
-. Because we have they  defined as `interface`, an user could create his own implementation, past it on the builder and later get an error that `ChatDomain` or `ChatClient` has not been initialized
-. On the "test side", if we are using a global instance instead of an injected one, we can't run our tests in parallel, or at least be sure our tests are working as we expect

I am not sure, but `ChatDomain` is a kind of wrapper of `ChatClient` and `Chat` a kind of wrapper of `ChatDomain`?

Maybe we can have a "common interface" (The same one that ChatClient uses), so the final user can use "whatever" implementation of `ChatClient` the want to use.
`ChatDomain` could implement it by delegating it to LLC, and `Chat` does the same, but delegating it on `ChatDomain`

Reply:
@JcMinarro common interface would be huge and not really needed. Top layers should simplify bottom ones by reducing amount of methods and narrowing down responsibilities:
```
chat: top layer which provided UI related api
domain: data layer (repository) which provides everything about data
client: network layer which provides raw network api
```

So UI components like messages list don't need `client` methods or know anything about it. Messages list should retrieve it from data layer.

Reply:
Ok, I agree that "Top Layers" should simplify "Bottom ones".
So we need that a user that is using "UI Chat", doesn't need to know anything about `ChatDomain` neither `ChatClient`.

I mean, a user should be able to do something like:
```
val chatUI = ChatUI.Builder(....).build() 
chatUI.sendMessage(...)
```

Without "Knowledge" of the bottom layer

Reply:
I opt for:

1. Hiding `ChatDomain` initialization into `Chat.Builder.build()`:
```
Chat.Builder.build() {
    // initialization of ChatDomain 
    ChatDomain.Builder().build()
    // building actual Chat instance
    // ....
}
```
2. Hiding `ChatClient` initialization into `ChatDomain.Builder.build()`:
```
ChatDomain.Builder.build() {
    // initialization of ChatClient 
    ChatClient.Builder().build()
    // building actual ChatDomain instance
    // ....
}
```
3. Adding `setUser()` function into `ChatDomain`, so that:
```
fun Chat.setUser(user: User) {
    ChatDomain.instance.setUser(user)
}
```
wdut?

Reply:
Yep

Reply:
Reviewed livedata library. Instance of `ChatDomain` deeply depends on current user, basically all internal components require current user. It is completely stateful and can't exists without current user. And that can't be refactored easily and fast.
Talked also with Thierry and it's done because database contains particular user related data. Which makes sense.

So for now it's probably better to create it when setUser is called. Later can be optimised. because deleting/creating database is async process. Async + stateful initialisation is tricky.
```kotlin
val chat = Chat.Builder(apiKey, context).build()
chat.setUser(user) // init client and domain
```